usr/lib/cca_stdll/cca_specific.c

```diff
@@ -1206,6 +1208,288 @@ CK_RV token_specific_rsa_verify(STDLL_TokData_t * tokdata,
+        goto done;
+    }
+
+    *((uint32_t *)message) = pss->sLen;
+    memcpy(message + 4, in_data, in_data_len);
+
+    /* The max value allowable by CCA for out_data_len is 512, so cap the
+     * incoming value if its too large. CCA will throw error 8, 72 otherwise.
+     */
+    if (out_data_len > 512)
+        out_data_len = 512;
+
+    rule_array_count = 2;
+    switch (pss->hashAlg) {
+    case CKM_SHA_1:
+        if (pss->mgf != CKG_MGF1_SHA1) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-1   ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA224:
+        if (pss->mgf != CKG_MGF1_SHA224) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-224 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA256:
+        if (pss->mgf != CKG_MGF1_SHA256) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-256 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA384:
+        if (pss->mgf != CKG_MGF1_SHA384) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-384 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    case CKM_SHA512:
+        if (pss->mgf != CKG_MGF1_SHA512) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCS-PSSSHA-512 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+    }
+
+    dll_CSNDDSV(&return_code,
+                &reason_code,
+                NULL,
+                NULL,
+                &rule_array_count,
+                rule_array,
+                (long *)&(attr->ulValueLen),
+                attr->pValue,
+                &message_len,
+                message,
+                (long *)&out_data_len, out_data);
+
+    if (return_code == 4 && reason_code == 429) {
+        rc = CKR_SIGNATURE_INVALID;
+        goto done;
+    } else if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSNDDSV (RSA PSS VERIFY) failed. return:%ld, reason:%ld\n",
+                    return_code, reason_code);
+        if (return_code == 8 && reason_code == 72) {
+            /*
+             * Return CKR_SIGNATURE_INVALID in case of return code 8 and
+             * reason code 72 because we dont know why the RSA op failed
+             * and it may have failed due to a tampered signature being
+             * greater or equal to the modulus.
+             */
+            rc = CKR_SIGNATURE_INVALID;
```

Comment:
can you add a debug or devel trace here ? May be of interest for a developer.

Reply:
you have the CCA return and reason codes traces already just above....
This is actually copied from token_specific_rsa_verify where it existed the same way already.